File: app/src/main/java/com/o4x/musical/util/ViewUtil.java
package com.o4x.musical.util;

import android.animation.Animator;
import android.animation.ArgbEvaluator;
import android.animation.ObjectAnimator;
import android.content.Context;
import android.content.res.ColorStateList;
import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.graphics.drawable.Drawable;
import android.graphics.drawable.RippleDrawable;
import android.graphics.drawable.ShapeDrawable;
import android.graphics.drawable.StateListDrawable;
import android.graphics.drawable.shapes.RectShape;
import android.os.Build;
import android.util.DisplayMetrics;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.PathInterpolator;
import android.widget.TextView;

import androidx.annotation.ColorInt;

import com.o4x.musical.R;
import com.simplecityapps.recyclerview_fastscroll.views.FastScrollRecyclerView;

import java.lang.reflect.Field;
import java.lang.reflect.Method;

import com.o4x.appthemehelper.util.ATHUtil;
import com.o4x.appthemehelper.util.ColorUtil;
import com.o4x.appthemehelper.util.MaterialValueHelper;

public class ViewUtil {

    public final static int PHONOGRAPH_ANIM_TIME = 400;

    public static Animator createBackgroundColorTransition(final View v, @ColorInt final int startColor, @ColorInt final int endColor) {
        return createColorAnimator(v, "backgroundColor", startColor, endColor);
    }

    public static Animator createTextColorTransition(final TextView v, @ColorInt final int startColor, @ColorInt final int endColor) {
        return createColorAnimator(v, "textColor", startColor, endColor);
    }

    private static Animator createColorAnimator(Object target, String propertyName, @ColorInt int startColor, @ColorInt int endColor) {
        ObjectAnimator animator;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            animator = ObjectAnimator.ofArgb(target, propertyName, startColor, endColor);
        } else {
            animator = ObjectAnimator.ofInt(target, propertyName, startColor, endColor);
            animator.setEvaluator(new ArgbEvaluator());
        }

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            animator.setInterpolator(new PathInterpolator(0.4f, 0f, 1f, 1f));
        }
        animator.setDuration(PHONOGRAPH_ANIM_TIME);
        return animator;
    }

    public static Drawable createSelectorDrawable(Context context, @ColorInt int color) {
        final StateListDrawable baseSelector = new StateListDrawable();
        baseSelector.addState(new int[]{android.R.attr.state_activated}, new ColorDrawable(color));

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            return new RippleDrawable(ColorStateList.valueOf(color), baseSelector, new ColorDrawable(Color.WHITE));
        }

        baseSelector.addState(new int[]{}, new ColorDrawable(Color.TRANSPARENT));
        baseSelector.addState(new int[]{android.R.attr.state_pressed}, new ColorDrawable(color));
        return baseSelector;
    }

    public static boolean hitTest(View v, View parent, int x, int y) {
        int[] viewLocation = new int[2];
        v.getLocationInWindow(viewLocation);

        int[] rootLocation = new int[2];
        parent.getLocationInWindow(rootLocation);

        final int vx = viewLocation[0] - rootLocation[0];
        final int vy  = viewLocation[1] - rootLocation[1];
        final int width = v.getWidth();
        final int height = v.getHeight();

        return (x >= vx) && (x <= vx + width) && (y >= vy) && (y <= vy + height);
    }

    public static void setUpFastScrollRecyclerViewColor(Context context, FastScrollRecyclerView recyclerView, int themeColor) {
        recyclerView.setPopupBgColor(themeColor);
        recyclerView.setPopupTextColor(MaterialValueHelper.getPrimaryTextColor(context, ColorUtil.INSTANCE.isColorLight(themeColor)));
        recyclerView.setThumbColor(themeColor);
        recyclerView.setTrackColor(ColorUtil.INSTANCE.withAlpha(ATHUtil.INSTANCE.resolveColor(context, R.attr.colorControlNormal), 0.12f));
    }

    public static float convertDpToPixel(float dp, Resources resources) {
        DisplayMetrics metrics = resources.getDisplayMetrics();
        return dp * metrics.density;
    }

    public static float convertPixelsToDp(float px, Resources resources) {
        DisplayMetrics metrics = resources.getDisplayMetrics();
        return px / metrics.density;
    }

    public static int getViewBackgroundColor(View view) {
        int color = Color.TRANSPARENT;
        Drawable background = view.getBackground();
        if (background instanceof ColorDrawable)
            color = ((ColorDrawable) background).getColor();

        return color;
    }

    public static Bitmap getBitmapFromView(View view) {
        Bitmap returnedBitmap = Bitmap.createBitmap(view.getWidth(), view.getHeight(),Bitmap.Config.ARGB_8888);
        Canvas canvas = new Canvas(returnedBitmap);
        Drawable bgDrawable =view.getBackground();
        if (bgDrawable!=null)
            bgDrawable.draw(canvas);
        else
            canvas.drawColor(Color.WHITE);
        view.draw(canvas);
        return returnedBitmap;
    }

    public static void setMargins (View v, int l, int t, int r, int b) {
        if (v.getLayoutParams() instanceof ViewGroup.MarginLayoutParams) {
            ViewGroup.MarginLayoutParams p = (ViewGroup.MarginLayoutParams) v.getLayoutParams();
            p.setMargins(l, t, r, b);
            v.requestLayout();
        }
    }

    public static void setScrollBarColor(View scr, @ColorInt int color) {
        try
        {
            Field mScrollCacheField = View.class.getDeclaredField("mScrollCache");
            mScrollCacheField.setAccessible(true);
            Object mScrollCache = mScrollCacheField.get(scr); // scr is your Scroll View

            Field scrollBarField = mScrollCache.getClass().getDeclaredField("scrollBar");
            scrollBarField.setAccessible(true);
            Object scrollBar = scrollBarField.get(mScrollCache);

            Method method = scrollBar.getClass().getDeclaredMethod("setVerticalThumbDrawable", Drawable.class);
            method.setAccessible(true);

            // Set your drawable here.
            method.invoke(scrollBar, new ColorDrawable(color));
        }
        catch(Exception e)
        {
            e.printStackTrace();
        }
    }
}

File: app/src/main/java/github/o4x/musical/util/ViewUtil.kt
package github.o4x.musical.util

import android.R
import android.animation.Animator
import android.animation.ArgbEvaluator
import android.animation.ObjectAnimator
import android.content.Context
import android.content.res.ColorStateList
import android.content.res.Resources
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.graphics.drawable.Drawable
import android.graphics.drawable.RippleDrawable
import android.graphics.drawable.StateListDrawable
import android.os.Build
import android.view.View
import android.view.ViewGroup.MarginLayoutParams
import android.view.animation.PathInterpolator
import android.widget.TextView
import androidx.annotation.ColorInt
import com.simplecityapps.recyclerview_fastscroll.views.FastScrollRecyclerView

object ViewUtil {
    const val PHONOGRAPH_ANIM_TIME: Int = 400

    fun createBackgroundColorTransition(
        v: View?,
        @ColorInt startColor: Int,
        @ColorInt endColor: Int
    ): Animator {
        return createColorAnimator(v, "backgroundColor", startColor, endColor)
    }

    fun createTextColorTransition(
        v: TextView?,
        @ColorInt startColor: Int,
        @ColorInt endColor: Int
    ): Animator {
        return createColorAnimator(v, "textColor", startColor, endColor)
    }

    private fun createColorAnimator(
        target: Any?,
        propertyName: String?,
        @ColorInt startColor: Int,
        @ColorInt endColor: Int
    ): Animator {
        val animator: ObjectAnimator
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            animator = ObjectAnimator.ofArgb(target, propertyName, startColor, endColor)
        } else {
            animator = ObjectAnimator.ofInt(target, propertyName, startColor, endColor)
            animator.setEvaluator(ArgbEvaluator())
        }

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            animator.setInterpolator(PathInterpolator(0.4f, 0f, 1f, 1f))
        }
        animator.setDuration(PHONOGRAPH_ANIM_TIME.toLong())
        return animator
    }

    fun createSelectorDrawable(context: Context?, @ColorInt color: Int): Drawable {
        val baseSelector = StateListDrawable()
        baseSelector.addState(intArrayOf(R.attr.state_activated), ColorDrawable(color))

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            return RippleDrawable(
                ColorStateList.valueOf(color),
                baseSelector,
                ColorDrawable(Color.WHITE)
            )
        }

        baseSelector.addState(intArrayOf(), ColorDrawable(Color.TRANSPARENT))
        baseSelector.addState(intArrayOf(R.attr.state_pressed), ColorDrawable(color))
        return baseSelector
    }

    @JvmStatic
    fun hitTest(v: View, parent: View, x: Int, y: Int): Boolean {
        val viewLocation = IntArray(2)
        v.getLocationInWindow(viewLocation)

        val rootLocation = IntArray(2)
        parent.getLocationInWindow(rootLocation)

        val vx = viewLocation[0] - rootLocation[0]
        val vy = viewLocation[1] - rootLocation[1]
        val width = v.getWidth()
        val height = v.getHeight()

        return (x >= vx) && (x <= vx + width) && (y >= vy) && (y <= vy + height)
    }

    fun setUpFastScrollRecyclerViewColor(
        context: Context,
        recyclerView: FastScrollRecyclerView,
        themeColor: Int
    ) {
        recyclerView.setPopupBgColor(themeColor)
        recyclerView.setPopupTextColor(context.textColorPrimary())
        recyclerView.setThumbColor(themeColor)
        recyclerView.setTrackColor(context.colorControlNormal().withAlpha(.12f))
    }

    fun convertDpToPixel(dp: Float, resources: Resources): Float {
        val metrics = resources.getDisplayMetrics()
        return dp * metrics.density
    }

    fun convertPixelsToDp(px: Float, resources: Resources): Float {
        val metrics = resources.getDisplayMetrics()
        return px / metrics.density
    }

    fun getViewBackgroundColor(view: View): Int {
        var color = Color.TRANSPARENT
        val background = view.getBackground()
        if (background is ColorDrawable) color = background.getColor()

        return color
    }

    fun getBitmapFromView(view: View): Bitmap {
        val returnedBitmap =
            Bitmap.createBitmap(view.getWidth(), view.getHeight(), Bitmap.Config.ARGB_8888)
        val canvas = Canvas(returnedBitmap)
        val bgDrawable = view.getBackground()
        if (bgDrawable != null) bgDrawable.draw(canvas)
        else canvas.drawColor(Color.WHITE)
        view.draw(canvas)
        return returnedBitmap
    }

    fun setMargins(v: View, l: Int, t: Int, r: Int, b: Int) {
        if (v.getLayoutParams() is MarginLayoutParams) {
            val p = v.getLayoutParams() as MarginLayoutParams
            p.setMargins(l, t, r, b)
            v.requestLayout()
        }
    }

    fun setScrollBarColor(scr: View?, @ColorInt color: Int) {
        try {
            val mScrollCacheField = View::class.java.getDeclaredField("mScrollCache")
            mScrollCacheField.setAccessible(true)
            val mScrollCache = mScrollCacheField.get(scr) // scr is your Scroll View

            val scrollBarField = mScrollCache!!.javaClass.getDeclaredField("scrollBar")
            scrollBarField.setAccessible(true)
            val scrollBar = scrollBarField.get(mScrollCache)

            val method = scrollBar!!.javaClass.getDeclaredMethod(
                "setVerticalThumbDrawable",
                Drawable::class.java
            )
            method.setAccessible(true)

            // Set your drawable here.
            method.invoke(scrollBar, ColorDrawable(color))
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }
}
